# META-WEPP: Metagenomic Wastewater-based Epidemiology using Phylogenetic Placement

## Table of Contents
- [Introduction](#intro)
- [Installation](#install)
  - [Option-1: Install via Conda](#dockerhub)
- [Quick Start](#example)

<br>


## <a name="intro"></a> Introduction


<div align="center">
    <img src="images/WEPP_Overview.png" width="600">
    <div><b>Figure 1: Overview of WEPP</b></div>
</div>
[Kraken2](https://github.com/DerrickWood/kraken2)
[MeSS](https://github.com/metagenlab/MeSS)
[WEPP](https://github.com/TurakhiaLab/WEPP)

META-WEPP is a Snakemake-based bioinformatics pipeline designed to enable rapid classification and haplotype-level analysis of mixed-pathogen metagenomic samples. Developed for flexible, high-throughput use in public health surveillance, META-WEPP integrates [Kraken2](https://github.com/DerrickWood/kraken2) for taxonomic classification and routes identified pathogen reads into [WEPP](https://github.com/TurakhiaLab/WEPP) for phylogenetic placement and haplotype inference. The pipeline automates the end-to-end workflow—from raw mixed reads to lineage-level analysis—with optional support for simulated read generation using [MeSS](https://github.com/metagenlab/MeSS). 


## <a name="install"></a> Installation

### <a name="dockerhub"></a> Option-1: Install via Conda.


**Step 0:**  Build a Kraken database
**1.** Install a taxonomy. 
```
kraken2-build --download-taxonomy --db $DBNAME
```
(Replace "$DBNAME" above with your preferred database name/location. The database will use approximately 100 GB of disk space during creation. )

**2.** Add sequence to the database's genomic library using the --add-to-library switch, e.g.:
```
kraken2-build --add-to-library /path/to/chr1.fa --db $DBNAME
kraken2-build --add-to-library /path/to/chr2.fa --db $DBNAME
```

Add a list of files to the database's genomic library
```
for file in /path/to/chr*.fa
do
    kraken2-build --add-to-library $file --db $DBNAME
done
```

(Optional) Install one or more reference libraries: https://github.com/DerrickWood/kraken2/wiki/Manual#standard-kraken-2-database
```
kraken2-build --download-library bacteria --db $DBNAME
```

**3.** Build the database 
```
kraken2-build --build --db $DBNAME
```
Customized kmer with `--kmer-len` and `--minimizer-len` option if needed.

**4.** (Optional) Remove intermediate files after building a custom database which helps to free disk space.
```
kraken2-build --clean --db $DBNAME
```
More information in https://github.com/DerrickWood/kraken2/wiki/Manual#custom-databases

**Step 1:** Clone the extension repository.
```bash
git clone https://github.com/TurakhiaLab/metagenomic-WBE.git
```
**Step 2:** Clone the WEPP repository in the extension repository.
```bash
cd metagenomic-WBE \
git clone https://github.com/TurakhiaLab/WEPP.git
```
**Step 3:** Install the conda environment package.
```bash
conda env create -f environment.yaml
conda activate meta-wepp-1
```

---

##  <a name="example"></a> Quick Start=

### <a name="rsv_a_example"></a> Example - 1: Run the pipeline with MeSS simulated data
**Step 1:** Build a Kraken database
```
// Go to the home directory or any directory to build kraken Database
mkdir kraken_DB
kraken2-build --download-taxonomy --db kraken_DB
kraken2-build --add-to-library /path/to/metagenomic-WBE/filtered_genomes.fa --db kraken_DB
kraken2-build --build --db kraken_DB
```
This will save the datasets on a separate data/RSVA_real folder within the repository.

**Step 2:**  Edit the `config.yaml` file

Put in your kraken database information information:
```
kraken_db: "/path/to/kraken_DB" # DBNAME Is the name of your database you created earlier
kraken_threads: 8 # You can change this to make it faster
kraken_report: "/path/to/{KRAKEN_REPORT_NAME}" # You can name this anything
kraken_output: "/path/to/{KRAKEN_OUTPUT_NAME}" # You can also name this to anything you want
...
simulation_tool: "MESS" # Set to "MeSS" to run the sim→merge→Kraken chain   
```

Input FASTA and MAT
```
REF: "filtered_genomes.fa" # The name of your FASTA file that you just downloaded (same as above, change to what they would download)
TREE: "{*.all.masked.pb.gz}" # The name of your MAT file you just downloaded (sane as above, change to what they would download)
```

The target taxid you want to analyze:
```
target_taxids:
  - 2697049 # The taxid (change to the taxid they would use)
```

**Step 3:**  Run the pipeline
```
snakemake --cores 32
```

**Step 4:**  Analyze Results

All results generated by WEPP can be found in the `WEPP/results/2697049` directory.

---

### <a name="arguments"></a> WEPP Arguments

From the WEPP repository:
> The WEPP Snakemake pipeline requires the following arguments, which can be provided either via the configuration file (`config/config.yaml`) or passed directly on the command line using the `--config` argument. The command line arguments take precedence over the config file.
> 1. `DIR` - Folder name containing the wastewater reads
> 2. `FILE_PREFIX` - File Prefix for all intermediate files 
> 3. `REF` - Reference Genome in fasta
> 4. `TREE` - Mutation-Annotated Tree
> 5. `SEQUENCING_TYPE` - Sequencing read type (s:Illumina single-ended, d:Illumina double-ended, or n:ONT long reads)
> 6. `PRIMER_BED` - BED file for primers from the `primers` folder
> 7. `MIN_AF` - Alleles with an allele frequency below this threshold in the reads will be masked. 
> 8. `MIN_Q` - Alleles with a Phred score below this threshold in the reads will be masked.
> 9. `MAX_READS` - Maximum number of reads considered by WEPP from the sample. Helpful for reducing runtime
> 10. `CLADE_IDX` - Index used for assigning clades to selected haplotypes from MAT. Generally '1' for SARS-CoV-2 MATs and '0' for others. > Could be checked by running: "matUtils summary -i {TREE} -C {FILENAME}" -> Use '0' for annotation_1 and '1' for annotation_2 

---

### <a name="example"></a> Usage
The entire pipeline can be ran through the following command:
```
snakemake --cores 32
```
This will run the full pipeline and run WEPP for each taxid in `target_taxids` in the `config.yaml` file (in the wepp extension directory).

